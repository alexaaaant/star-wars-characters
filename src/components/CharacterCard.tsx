import { Card, CardContent, Grid, Typography } from '@mui/material';
import { observer } from 'mobx-react-lite';
import React from 'react';
import { Link } from 'react-router-dom';
import { Character } from '../types/characters';

interface CharacterCardProps {
  character: Character;
  observerRef: ((el: HTMLDivElement) => void) | null;
}

export const CharacterCard: React.FC<CharacterCardProps> = observer(
  ({ character, observerRef }) => {

    return (
      <>
        {character ? (
          <Grid item={true} xs={12} sm={6} md={4} key={character.name}>
            <Link
              to={`/character/${character.name}`}
              style={{ textDecoration: 'none' }}
            >
              <Card>
                <CardContent>
                  <Typography variant="h5" component="h2">
                    {character.name}
                  </Typography>
                </CardContent>
              </Card>
            </Link>
            {observerRef && <div ref={observerRef} />}
          </Grid>
        ) : (
          ''
        )}
      </>
    );
  }
);

CharacterCard.displayName = 'CharacterCard';
