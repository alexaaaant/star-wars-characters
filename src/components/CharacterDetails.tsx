export function CharacterDetails() {
  return <div>Details</div>;
}
