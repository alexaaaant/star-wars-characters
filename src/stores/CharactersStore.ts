import { Character } from '../types/characters';
import { makeObservable, observable, action, runInAction } from 'mobx';

interface GetCharactersResp {
    results: Character[];
    next: string;
}

interface ReqParams {
    page: number;
    search: string;
}

class CharactersStore {
    characters: Map<string, Character> = new Map();
    reqParams: ReqParams = { page: 1, search: '' };
    completed = false;
    isLoading = false;
    
    constructor() {
        makeObservable(this, {
            characters: observable, 
            completed: observable,
            isLoading: observable,

            getCharacters: action,
            loadNextPage: action,
            startSearch: action,
            updateCharacter: action,
        });
    }

    getCharacters = async () => {
        try {
            if (this.completed) {
                return;
            }
            this.isLoading = true;
            const { page, search } = this.reqParams;
            const response = await fetch(`https://swapi.dev/api/people?page=${page}&search=${search}`);
            const data: GetCharactersResp = await response.json();
            runInAction(() => {
                if (data.results) {
                    data.results.forEach((character) => {
                        this.characters.set(character.name, character);
                    });
                }
                if (!data.next) {
                    this.completed = true;
                }
            });
          } catch (error) {
            console.log('Error loading characters:', error);
            return null;
          } finally {
            runInAction(() => {
                this.isLoading = false;
            });
          }
    }

    startSearch = (request: ReqParams['search']) => {
        if (this.reqParams.search !== request) {
            this.characters.clear();
            this.completed = false;
            this.reqParams.page = 1;
        }
        this.reqParams.search = request;
        this.getCharacters();
    }

    loadNextPage = () => {
        if (!this.completed) {
            this.reqParams.page++;
            this.getCharacters();
        }
    }

    updateCharacter = (name: Character['name'], props: Partial<Character>) => {
        const character = this.characters.get(name)!;
        
        Object.keys(props).forEach((key) => {
            const typedKey = key as keyof  Character;
            character[typedKey] = props[typedKey]!;
        });
    }
}

export const charactersStore = new CharactersStore();
