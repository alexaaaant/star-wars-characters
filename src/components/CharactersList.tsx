import { useState, useEffect, useCallback, useRef, ReactNode } from 'react';
import {
  CircularProgress,
  Container,
  Grid,
  TextField,
  Typography,
} from '@mui/material';
import { useDebounce } from '../common/hooks/useDebounce';
import { observer } from 'mobx-react-lite';
import { charactersStore } from '../stores/CharactersStore';
import { CharacterCard } from './CharacterCard';

export const CharactersList = observer(() => {
  const [searchQuery, setSearchQuery] = useState('');
  const debouncedSearchQuery = useDebounce(searchQuery);
  const observerComp = useRef<IntersectionObserver | null>(null);
  const initial = useRef(true);
  const { characters, startSearch, loadNextPage, isLoading } =
    charactersStore;

  useEffect(() => {
    if (initial.current && characters.size) {
      return;
    }
    initial.current = false;
    startSearch(debouncedSearchQuery);
  },        [debouncedSearchQuery, startSearch]);

  const observerRef = useCallback(
    (node: HTMLDivElement) => {
      if (!observerComp.current) {
        observerComp.current = new IntersectionObserver((entries) => {
          if (entries[0].isIntersecting) {
            loadNextPage();
          }
        });
      }

      if (node !== null) {
        observerComp.current.observe(node);
      } else {
        observerComp.current.disconnect();
      }
    },
    [loadNextPage]
  );

  const handleSearchChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setSearchQuery(event.target.value);
  };

  const charactersElements = () => {
    const elements: ReactNode[] = [];
    let i = 0;
    characters.forEach((character) => {
      elements.push(
        <CharacterCard
          key={character.name}
          character={character}
          observerRef={i === characters.size - 1 ? observerRef : null}
        />
      );
      i++;
    });
    return elements;
  };

  return (
    <Container>
      <TextField
        label="Search"
        value={searchQuery}
        onChange={handleSearchChange}
        sx={{ m: 2 }}
      />
      {!!characters.size && (
        <Grid container={true} spacing={2}>
          {charactersElements()}
        </Grid>
      )}
      <Grid container={true} justifyContent={'center'}>
        {!characters.size && !isLoading && (
          <Typography variant="h5" component="h2">
            Empty
          </Typography>
        )}
        {isLoading && <CircularProgress sx={{ m: 2 }} />}
      </Grid>
    </Container>
  );
});

CharactersList.displayName = 'CharactersList';
