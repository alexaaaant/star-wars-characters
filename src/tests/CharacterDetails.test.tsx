import { render, screen, fireEvent } from '@testing-library/react';
import { CharacterDetails } from '../components/CharacterDetails';
import { charactersStore } from '../stores/CharactersStore';
import '@testing-library/jest-dom';

jest.mock('react-router-dom', () => ({
  useParams: () => ({ id: 'Luke Skywalker' }),
}));

// jest.mock('../stores/CharactersStore', () => ({
//   charactersStore: {
//     characters: new Map([
//       ['Luke Skywalker', { name: 'Luke Skywalker', birth_year: '19BBY', gender: 'male', height: '172' }],
//       ['Leia Organa', { name: 'Leia Organa', birth_year: '19BBY', gender: 'female', height: '150' }],
//     ]),
//     updateCharacter: jest.fn(),
//   },
// }));

describe('CharacterDetails', () => {
  beforeEach(() => {
    charactersStore.characters = new Map([
      [
        'Luke Skywalker',
        {
          name: 'Luke Skywalker',
          birth_year: '19BBY',
          gender: 'male',
          height: '172',
        },
      ],
      [
        'Leia Organa',
        {
          name: 'Leia Organa',
          birth_year: '19BBY',
          gender: 'female',
          height: '150',
        },
      ],
    ]);
  });

  it('makes a snapshot', () => {
    const { container } = render(<CharacterDetails />);

    expect(container).toMatchSnapshot();
  });

  it('renders character details', () => {
    render(<CharacterDetails />);

    const nameElement = screen.getByText('Luke Skywalker');
    const birthYearInput = screen.getByLabelText('Birth Year');
    const genderInput = screen.getByLabelText('Gender');
    const heightInput = screen.getByLabelText('Height');

    expect(nameElement).toBeInTheDocument();
    expect(birthYearInput).toHaveValue('19BBY');
    expect(genderInput).toHaveValue('male');
    expect(heightInput).toHaveValue('172');
  });

  it('updates character birth year', () => {
    render(<CharacterDetails />);

    const birthYearInput = screen.getByLabelText('Birth Year');

    const newValue = '20BBY';
    fireEvent.change(birthYearInput, { target: { value: newValue } });

    expect(birthYearInput).toHaveValue(newValue);
  });

  it('updates character gender', () => {
    render(<CharacterDetails />);

    const genderInput = screen.getByLabelText('Gender');

    const newValue = 'female';
    fireEvent.change(genderInput, { target: { value: newValue } });

    expect(genderInput).toHaveValue(newValue);
  });

  it('updates character height', () => {
    render(<CharacterDetails />);

    const heightInput = screen.getByLabelText('Height');

    const newValue = 'female';
    fireEvent.change(heightInput, { target: { value: newValue } });

    expect(heightInput).toHaveValue(newValue);
  });
});
