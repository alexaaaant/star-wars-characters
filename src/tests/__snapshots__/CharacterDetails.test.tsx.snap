// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterDetails makes a snapshot 1`] = `
<div>
  <div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 css-1fyyp8j-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <h2
              class="MuiTypography-root MuiTypography-h5 css-1pe23e4-MuiTypography-root"
            >
              Luke Skywalker
            </h2>
            <div
              class="MuiFormControl-root MuiTextField-root css-yfcmxe-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="true"
                for=":r0:"
                id=":r0:-label"
              >
                Birth Year
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r0:"
                  type="text"
                  value="19BBY"
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-14lo706"
                  >
                    <span>
                      Birth Year
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiTextField-root css-yfcmxe-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="true"
                for=":r1:"
                id=":r1:-label"
              >
                Gender
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r1:"
                  type="text"
                  value="male"
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-14lo706"
                  >
                    <span>
                      Gender
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiTextField-root css-yfcmxe-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="true"
                for=":r2:"
                id=":r2:-label"
              >
                Height
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r2:"
                  type="text"
                  value="172"
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-14lo706"
                  >
                    <span>
                      Height
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
